File: Documentation/5.4/Samples/python/Indexes/IndexingPolymorphicData.py
from abc import ABC
from typing import Type

from ravendb import DocumentStore
from ravendb.documents.conventions import DocumentConventions
from ravendb.documents.indexes.abstract_index_creation_tasks import AbstractMultiMapIndexCreationTask

from examples_base import ExampleBase


class Animal(ABC):
    def __init__(self, name: str = None):
        self.name = name


class Dog(Animal): ...


class Cat(Animal): ...


# region multi_map_1
class Animals_ByName(AbstractMultiMapIndexCreationTask):
    def __init__(self):
        super().__init__()
        self._add_map("from c in docs.Cats select new { c.name }")
        self._add_map("from d in docs.Dogs select new { d.name }")


# endregion


class MultiMapIndexes(ExampleBase):
    def setUp(self):
        super().setUp()

    def test_multi_map_indexes(self):
        with self.embedded_server.get_document_store("MultiMapIndexes") as store:
            with store.open_session() as session:
                Animals_ByName().execute(store)
                session.store(Cat("Mitzy"))
                session.store(Dog("Mitzy"))
                session.save_changes()
                # region multi_map_2
                results = list(session.query_index_type(Animals_ByName, Animal).where_equals("name", "Mitzy"))
                # endregion

        # region other_ways_1
        store = DocumentStore()

        def _custom_find_collection_name(object_type: Type) -> str:
            if issubclass(object_type, Animal):
                return "Animals"
            return DocumentConventions.default_get_collection_name(object_type)

        store.conventions.find_collection_name = _custom_find_collection_name
        # endregion

File: Documentation/5.4/Samples/php/Indexes/IndexingPolymorphicData.php
<?php

namespace RavenDB\Samples\Indexes;

use RavenDB\Documents\Conventions\DocumentConventions;
use RavenDB\Documents\DocumentStore;
use RavenDB\Documents\Indexes\AbstractMultiMapIndexCreationTask;

class IndexingPolymorphicData
{
    public function MultiMapIndexes(): void
    {
        $store = new DocumentStore();
        try {
            $session = $store->openSession();
            try {
                # region multi_map_2
                /** @var array<Animal> $results */
                $results = $session
                    ->advanced()
                    ->documentQuery(Animal::class, Animals_ByName::class)
                    ->whereEquals("Name", "Mitzy")
                    ->toList();
                # endregion
            } finally {
                $session->close();
            }

            $session = $store->openSession();
            try {
                # region multi_map_3
                /** @var array<Animal> $results */
                $results = $session
                    ->query(Animal::class, Animals_ByName::class)
                    ->whereEquals("Name", "Mitzy")
                    ->toList();
                # endregion
            } finally {
                $session->close();
            }
        } finally {
            $store->close();
        }
    }
}

# region multi_map_1
class Animals_ByName extends AbstractMultiMapIndexCreationTask
{
    public function __construct()
    {
        parent::__construct();

        $this->addMap("from c in docs.Cats select new { c.name }");
        $this->addMap("from d in docs.Dogs select new { d.name }");
    }
}
# endregion

class OtherWays
{
    public function sample(): void
    {
        # region other_ways_1
        $store = new DocumentStore();
        $store->getConventions()->setFindCollectionName(
            function (?string $className): string {
                if (is_a($className, Animal::class)) {
                    return "Animals";
                }
                return DocumentConventions::defaultGetCollectionName($className);
            }
        );
        # endregion
    }
}

class Animal
{
    private ?string $name = null;

    public function getName(): ?string
    {
        return $this->name;
    }

    public function setName(?string $name): void
    {
        $this->name = $name;
    }
}

class Cat extends Animal
{
}

class Dog extends Animal
{
}
